File: MCP_GATEWAY/SKill_SOCKET-Mcp_gateway/agents/perplexityAgent.js
class PerplexityAgent{
    constructor(llmClient,toolRegistry){
        this.llmClient=llmClient;
        this.websearchTool=toolRegistry.get('web_search');
    
    }
    async run(query){
        const searchResults=await this.websearchTool(query);
        if (!searchResults?.results?.length) {
            return { answer: "Sorry, I couldn't find relevant information online.", sources: [] };
        }
         const context = searchResults.results.map((res, i) => `Source [${i + 1}]: ${res.content} (URL: ${res.url})`).join('\n\n');
        const prompt = `User's question: "${query}"\n\nContext:\n${context}\n\nBased only on the provided context, write a comprehensive answer. Cite sources using the format [1], [2], etc.`;
        const answer = await this.llmClient.generateText(prompt);
        const sources = searchResults.results.map(res => ({ url: res.url, title: res.title }));
        return { answer, sources: [...new Map(sources.map(item => [item.url, item])).values()] };
    }
}
module.exports=PerplexityAgent;

File: MCP_GATEWAY/SKill_SOCKET-Mcp_gateway/agents/perplexity_agent.py
class PerplexityAgent:
    """Answers questions using web search results + LLM synthesis."""

    def __init__(self, llm_client, tool_registry):
        self.llm_client = llm_client
        self.websearch_tool = tool_registry.get("web_search")

    async def run(self, query: str) -> dict:
        search_results = await self.websearch_tool(query)

        results = search_results.get("results", [])
        if not results:
            return {
                "answer": "Sorry, I couldn't find relevant information online.",
                "sources": [],
            }

        context = "\n\n".join(
            f"Source [{i + 1}]: {r['content']} (URL: {r['url']})"
            for i, r in enumerate(results)
        )

        prompt = (
            f'User\'s question: "{query}"\n\n'
            f"Context:\n{context}\n\n"
            "Based only on the provided context, write a comprehensive answer. "
            "Cite sources using the format [1], [2], etc."
        )

        answer = await self.llm_client.generate_text(prompt)

        # Deduplicate sources by URL
        seen: dict[str, dict] = {}
        for r in results:
            if r["url"] not in seen:
                seen[r["url"]] = {"url": r["url"], "title": r.get("title", "")}

        return {"answer": answer, "sources": list(seen.values())}
